openpifpaf/export_onnx.py

```diff
@@ -87,6 +87,41 @@ def apply(model, outfile, verbose=True):
     )
 
 
+def apply_components(model, outfile, input_w=129, input_h=97, verbose=True):
+    output_files = []
+    torch.onnx.export(
+        model.base_net,
+        torch.randn(1, 3, input_h, input_w),
+        outfile + '.base.onnx', verbose=verbose,
+        input_names=['input_batch'], output_names=['features'],
+        keep_initializers_as_inputs=True,
+        # opset_version=10,
+        do_constant_folding=True,
+        export_params=True,
+    )
+    output_files.append(outfile + '.base.onnx')
+
+    for head_i, head in enumerate(model.head_nets):
```

Comment:
Could we put all the heads in a single exported networks?
Having a separate net for each head complicates a lot postprocessing as we currently  have no way to have one postprocessing code taking output from two networks.

Reply:
But that is what we had before no? Without the breakdown by components? Maybe I am still misunderstanding what you need.

Reply:
Ah sorry, I try to explain better.
The ideal would be to split the network in two parts:
- base (contains the backbone)
- heads (contains all the heads)

Having a separate network for each head could in principle allow to parallelize the execution, but they are so small the the benefit is negligible, so I think it is better to keep all the heads together as this allows to avoid changes in the structure of tje postprocessing code.

Reply:
Why split the base and the head then?

Reply:
The base is the most compute-intensive part of the network. This is the part that is definitely worth running on accelerated HW.
The heads are quite small but contains "tricky" operators that are often not supported well or not supported at all by HW accelerators. Splitting the network in two allows to run the backbone on accelerated HW, while the heads part can be implemented in SW so that we have more flexibility in supporting the required node types.

Reply:
Before we had an in-place add operation that created an ATen op that wasn't supported but that is fixed. Otherwise the head should be the simplest part of the network: a convolution, some slicing and some constant offsets to add. Do you have a specific op that is still "tricky"?

Reply:
I will check better tomorrow but if I remember correctly Gather node for example is not supported well in tensor-rt

Reply:
Interesting. Didn't know that. Thanks. Have a good evening.

Reply:
Any new insight into which operator is the problematic one ... or two?

Reply:
Hi Sven thanks, we are double-checking this,  I will update here when we have more info.

Reply:
Looks like Slice layer with negative index is not supported:

```
(Unnamed Layer* 175) [Slice]: slice is out of input range
While parsing node number 178 [Slice]:
ERROR: onnx2trt_utils.hpp:347 In function convert_axis:
[8] Assertion failed: axis >= 0 && axis < nbDims
[2020-07-16 12:07:42.362] [error] error: could not parse onnx engine
[2020-07-16 12:07:42.363] [error] error: could not deserialize or build engine

```
In effect "ends" is -1. This seems a simple change, if this is something you can do I can retry quickly.

Reply:
This seems to be messed up in ONNX TensorRT.

Regardless, PyTorch also tried to be smart about intermediate shapes. Although the upsampled feature map sizes are always static, it tried to obtain them from the graph leading to lots of extra nodes in ONNX. I just pushed a commit that removes a lot of nodes from the head networks in ONNX. The heads are fairly simple now. The shapes should all be static positive numbers now. Let me know how that goes.

![image](https://user-images.githubusercontent.com/2775078/87679893-095ab380-c77d-11ea-8576-1b0313bb639c.png)

Reply:
Thanks, we will give this a try and let you know!

Reply:
It works now?

Reply:
Hi Sven, yes model import works correctly now in tensor-rt. Unfortunately inference execution fails, we are now checking if this is an issue of the model itself or there is some issue in the way we use tensor-rt. I'll update you ASAP
Thanks!

Reply:
So.. model import works correctly in tensorrt, but inference fails. We get this error message:

```
NVMEDIA_DLA :  308, ERROR: setInputTensorDesc failed 
NVMEDIA_DLA :  446, ERROR: SetInputTensorDesc failed for tensor: 7. status: 0x0.
NVMEDIA_DLA : 1914, ERROR: BindInputTensorArgs failed (Input). status: 0x7.
../rtExt/dla/native/dlaUtils.cpp (212) - DLA Error in submit: 7 (Failure to submit program to DLA engine.)
FAILED_EXECUTION: std::exception

```
If we keep our code the same but use just the "head" part of the network, then the inference also completes successfully (I still have to analye the resulting tensor to be sure it contains what we expect).

We are still trying to understand what is going wrong, but from what we see a split of the network in two parts (base + head as discussed) would allow us to run the (heavy) base in tensorrt and the rest in LPDNN right now.

For your reference i attach the log result in both cases.

Reply:
For some reason file upload fails.
I copy here the success case (running the base only):

```
Detected 1 inputs and 1 output network tensors.
[2020-07-21 17:14:34.121] [info] serialize engine to 
[2020-07-21 17:14:34.121] [warning] empty engine file name, skip save
[2020-07-21 17:14:34.125] [info] create execute context and malloc device memory...
[2020-07-21 17:14:34.125] [info] init engine...
[2020-07-21 17:14:34.137] [info] malloc device memory
nbBingdings: 2
[2020-07-21 17:14:34.138] [info] input: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 0, name: input_batch, size in byte: 150156
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
3 x 97 x 129   
[2020-07-21 17:14:34.138] [info] output: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 1, name: features, size in byte: 516096
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
2048 x 7 x 9   
Binding dims (size) for input_batch ind 0 is {3, 97, 129} (150156)
Binding dims (size) for features ind 1 is {2048, 7, 9} (516096)
[2020-07-21 17:14:34.150] [info] net forward takes 9.46506 ms

```

Reply:
And here the failure case (running the entire network):

```
Detected 1 inputs and 1 output network tensors.
[2020-07-21 17:14:34.121] [info] serialize engine to 
[2020-07-21 17:14:34.121] [warning] empty engine file name, skip save
[2020-07-21 17:14:34.125] [info] create execute context and malloc device memory...
[2020-07-21 17:14:34.125] [info] init engine...
[2020-07-21 17:14:34.137] [info] malloc device memory
nbBingdings: 2
[2020-07-21 17:14:34.138] [info] input: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 0, name: input_batch, size in byte: 150156
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
3 x 97 x 129   
[2020-07-21 17:14:34.138] [info] output: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 1, name: features, size in byte: 516096
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
2048 x 7 x 9   
Binding dims (size) for input_batch ind 0 is {3, 97, 129} (150156)
Binding dims (size) for features ind 1 is {2048, 7, 9} (516096)
[2020-07-21 17:14:34.150] [info] net forward takes 9.46506 ms

```

Reply:
Failure and success seem to have identical output. Was that a copy&paste fail?